server/oauth_migration.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-api/cluster"
+	"github.com/pkg/errors"
+)
+
+const (
+	oauthMigrationStoreKey = "oauth_migration"
+	oauthMigrationMutexKey = "oauth_migration_mutex"
+
+	oauthMigrationStatusInProgress = "IN_PROGRESS"
+	oauthMigrationStatusComplete   = "COMPLETE"
+	oauthMigrationStatusError      = "ERROR"
+)
+
+func (p *Plugin) checkAndPerformOAuthTokenMigration() error {
+	mutexAPI := cluster.MutexPluginAPI(p.API)
+	mutex, err := cluster.NewMutex(mutexAPI, oauthMigrationMutexKey)
+	if err != nil {
+		return err
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var status string
+	err = p.client.KV.Get(oauthMigrationStoreKey, &status)
+	if err != nil {
+		return err
+	}
+
+	// Migration is in progress or already completed
+	if status != "" {
+		return nil
+	}
+
+	_, err = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusInProgress)
+	if err != nil {
+		p.client.Log.Warn("error setting migration status to "+oauthMigrationStatusInProgress, "error", err.Error())
+	}
+
+	err = p.notifyAllConnectedUsersToReconnect()
+	if err != nil {
```

Comment:
(question) `notifyAllConnectedUsersToReconnect` returns an error only on listing failing. Since we're counting the errors per user, should this also fail if `numErrors` matches `allKeys` length (ie it failed for every single user), as it may indicate some issue with the migration? Or would it be an overkill/edge case?